axelrod/strategies/adaptivezerodet.py

```diff
@@ -0,0 +1,120 @@
+		game = self.match_attributes["game"]
+		if len(self.history):
+			last_round = (self.history[-1], opponent.history[-1])
+			scores = game.score(last_round)
+			self.scores[last_round[0]] += scores[0]
+
+	def strategy(self, opponent: Player) -> Action:
+		s = self.s
+		phi = self.phi
+		l = self.l
+		d = randint(0, 9)/1000 # Selects random value to adjust s and l
+		if self.scores[C] > self.scores[D] & len(self.history):
+			# This checks scores to determine how to adjust s and l either up or down by d
+			# This also checks if the length of the game is long enough to start adjusting
+			self.l = l+d
+			l = self.l
+			# adjust l up
+			self.s = s-d
+			s = self.s
+			# adjust s down
+			R, P, S, T = self.match_attributes["game"].RPST()
+			phi = self.phi
+			s_min = - min((T - l) / (l - S), (l - S) / (T - l)) # Sets minimum for s
+			if  (l > R) or (s < s_min):
+				# This checks that neither s nor l is leaving its range
+				if (l > R):
+					l = l-d
+					self.l = (l+R)/2
+					l = self.l
+					# If l would leave its range instead its distance from its max is halved
+				if (s < s_min):
+					s = s+d
+					self.s = (s+s_min)/2
+					s = self.s
+					# If s would leave its range instead its distance from its min is halved
+			p1 = 1 - phi * (1 - s) * (R - l)
+			p2 = 1 - phi * (s * (l - S) + (T - l))
+			p3 = phi * ((l - S) + s * (T - l))
+			p4 = phi * (1 - s) * (l - P)
+			four_vector = [p1, p2, p3, p4]
+			# Four vector is calculated with new parameters
+			self.set_four_vector(four_vector)
+			if not hasattr(self, "_four_vector"):
+				raise ValueError("_four_vector not yet set")
+			if len(opponent.history) == 0:
+				return self._initial
+			p = self._four_vector[(self.history[-1], opponent.history[-1])]
+			return random_choice(p)
+		else:
+			# This adjusts s and l in the opposite direction
+			self.l = l-d
+			l = self.l
+			# adjust l down
+			self.s = s+d
+			s = self.s
+			# adjust s up
+			R, P, S, T = self.match_attributes["game"].RPST()
+			phi = self.phi
+			if (l < P) or (s > 1):
+				# This checks that neither s nor l is leaving its range
+				if (l < P):
+					l = l+d
+					self.l = (l+P)/2
+					l = self.l
+					# If l would leave its range instead its distance from its min is halved
+				if (s > 1):
+					s = s-d
+					self.s = (s+1)/2
+					s = self.s
+					# If s would leave its range instead its distance from its max is halved
+			p1 = 1 - phi * (1 - s) * (R - l)
+			p2 = 1 - phi * (s * (l - S) + (T - l))
+			p3 = phi * ((l - S) + s * (T - l))
+			p4 = phi * (1 - s) * (l - P)
+			four_vector = [p1, p2, p3, p4]
+			# Four vector is calculated with new parameters
+			self.set_four_vector(four_vector)
+			if not hasattr(self, "_four_vector"):
+				raise ValueError("_four_vector not yet set")
+			if len(opponent.history) == 0:
+				return self._initial
+			p = self._four_vector[(self.history[-1], opponent.history[-1])]
+			return random_choice(p)

```

Comment:
The code in the two branches of this if/else is very similar. Could it perhaps be refactored into a function controlled by a parameter or two?

Reply:
refactored